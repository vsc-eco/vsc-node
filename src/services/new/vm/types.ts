import { ContractErrorType } from "./utils.js";

type Message = {type: string};

export type Env = {
  'anchor.id': string
  'anchor.block': string
  'anchor.timestamp': number
  'anchor.height': number
  'msg.sender': string
  'msg.required_auths': string[]
  'tx.origin': string
}

export type PartialResultMessage = Message & {
    type: 'partial-result',
    contract_id: string,
    index: string, // TODO change to number
    stateMerkle: string,
};

export type FinishResultMessage = Message & {
    type: 'finish-result'
}

export type ReadyMessage = Message & {
    type: 'ready'
}

export type ExecuteStopMessage = Message & {
    type: 'execute-stop',
    ret: string | null,
    errorType: ContractErrorType | null,
    error: string | null | {
        msg: string,
        file: string,
        line: number,
        colm: number,
    }
    logs:( string | number | boolean)[],
    IOGas: number,
    reqId: string,
    ledger: EventOp[],
}

export type CallMessage = Message & {
    type: 'call'
    contract_id: string,
    payload: string,
    action: string,
    intents: string[],
    balance_map: Record<string, {
        HBD: number
        HIVE: number
    }>
    env: Env,
    reqId: string,
}

export type FinishMessage = Message & {
    type: 'finish'
}

export type AnySentMessage = CallMessage | FinishMessage

export type AnyReceivedMessage = PartialResultMessage | FinishResultMessage | ReadyMessage | ExecuteStopMessage


export enum EventOpType {
    'ledger:transfer' = 110_001,
    'ledger:withdraw' = 110_002,
    'ledger:deposit' = 110_003,
  
    //Reserved for future, DO NOT USE
    'ledger:stake_hbd' = 110_004,
    'ledger:unstake_hbd' = 110_005,
    'ledger:claim_hbd' = 110_006,
    
    //Reserved for future, DO NOT USE
    'consensus:stake' = 100_001,
    'consensus:unstake' = 100_002
    
  }
  
  
  export interface EventOp {
    owner: string
    tk: 'HBD' | 'HIVE'
    t: EventOpType
    amt: number
    memo?: string
    
  }