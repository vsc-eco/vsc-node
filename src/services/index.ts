import * as IPFS from "ipfs-http-client";
import { IPFSHTTPClient } from "ipfs-http-client";
import Path from 'path'
import os from 'os'
import Crypto from 'crypto'
import { Config } from "./nodeConfig";
import { Ed25519Provider } from "key-did-provider-ed25519";
import { DID } from "dids";
import KeyResolver from 'key-did-resolver'
import { TransactionPoolService } from "./transactionPool";
import { mongo } from "./db";
import { Db } from "mongodb";
import { ChainBridge } from "./chainBridge";
//@ts-ignore
import type { CeramicClient } from "@ceramicnetwork/http-client";
import { ContractEngine } from "./contractEngine";
import { P2PService } from "./pubsub";

interface CoreOptions {
    pathSuffix?: string
    dbSuffix?: string
    ipfsApi?: string
}

export class CoreService {
    ipfs: IPFSHTTPClient;
    config: Config;
    identity: DID;
    wallet: DID;
    transactionPool: TransactionPoolService;
    db: Db;
    chainBridge: ChainBridge;
    ceramic: CeramicClient;
    contractEngine: ContractEngine;
    options: CoreOptions;
    p2pService: P2PService;

    constructor(options?: CoreOptions) {
        this.options = options || {};

        if(!this.options.ipfsApi) {
            this.options.ipfsApi = "/ip4/127.0.0.1/tcp/5001"
        }
    }

    async start() {
        this.ipfs = IPFS.create(this.options.ipfsApi);
        const homeDir = this.options.pathSuffix ? Path.join(os.homedir(), '.vsc-node-' + this.options.pathSuffix) : Path.join(os.homedir(), '.vsc-node')
        this.config = new Config(homeDir)
        await this.config.open()
        this.db = this.options.dbSuffix ? mongo.db('vsc-' + this.options.dbSuffix) : mongo.db('vsc')
        await mongo.connect()

        for(let key of ['node', 'wallet']) {
            let privateKey = null
            if (this.config.get(`identity.${key}Private`)) {
              privateKey = Buffer.from(this.config.get(`identity.${key}Private`), 'base64')
            } else {
              privateKey = Crypto.randomBytes(32)
              const hex = privateKey.toString('base64')
              this.config.set(`identity.${key}Private`, hex)
            }
            const keyPrivate = new Ed25519Provider(privateKey)
            const did = new DID({ provider: keyPrivate, resolver: KeyResolver.getResolver() })
            await did.authenticate()
            this.config.set(`identity.${key}Public`, did.id)
            if(key === "node") {
                this.identity = did
            }
            if(key === "wallet") {
                this.wallet = did;
            }
        }

        try 
        {
            const {CeramicClient} = await import('@ceramicnetwork/http-client')
            this.ceramic = new CeramicClient('https://ceramic.web3telekom.xyz')
            await this.ceramic.setDID(this.wallet)
    
            this.transactionPool = new TransactionPoolService(this)
            await this.transactionPool.start()
            
            this.chainBridge = new ChainBridge(this)
            await this.chainBridge.start();
    
            this.contractEngine = new ContractEngine(this)
            await this.contractEngine.start()
    
            this.p2pService = new P2PService(this)
            await this.p2pService.start()

        }
        catch (err) {
            console.trace(err)
        }
    }
}