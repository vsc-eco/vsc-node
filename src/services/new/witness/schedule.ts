import networks from '../../../services/networks'
import seedrandom from 'seedrandom'
import type { WithId } from 'mongodb'
import type { ElectionResult } from './electionManager'
import type { NewCoreService } from '..'
import type { WitnessServiceV2 } from '.'

function seedRand(func: () => number, min: number, max: number): number {
  return Math.floor(func() * (max - min + 1)) + min
}

function shuffle<T>(arr: T[], seed: string): T[] {
  const size = arr.length
  const rng = seedrandom(seed)
  const resp: T[] = []
  const keys: number[] = []

  for (let i = 0; i < size; i++) keys.push(i)
  for (let i = 0; i < size; i++) {
    const r = seedRand(rng, 0, keys.length - 1)
    const g = keys[r]
    keys.splice(r, 1)
    resp.push(arr[g])
  }
  return resp
}

type WithScheduleBN<T> = T & {
  bn: number
  bn_works: boolean
  in_past: boolean
}

/**
 * Applies block numbers to witness schedule
 */
function applyBNSchedule<T>(
  schedule: T[],
  network: NetworkInfo,
  consensusRound: ConsensusRound,
): WithScheduleBN<T>[] {
  const { roundLength } = network

  return schedule.map((e, index) => {
    return {
      ...e,
      bn: consensusRound.pastRoundHeight + index * roundLength,
      bn_works: (consensusRound.pastRoundHeight + index * roundLength) % roundLength === 0,
      in_past:
        consensusRound.pastRoundHeight + index * roundLength < consensusRound.currentBlockNumber,
    }
  })
}

function weightedSchedule(
  network: NetworkInfo,
  witnessNodes: {
    account: string
    key: string
  }[],
  electionResult: WithId<ElectionResult>,
  randomizeHash: string,
  consensusRound: ConsensusRound,
) {
  const { totalRounds } = network

  const outSchedule: {
    account: string
    key: string
  }[] = []
  for (let x = 0; x < totalRounds; x++) {
    if (witnessNodes[x % witnessNodes.length]) {
      outSchedule.push(witnessNodes[x % witnessNodes.length])
    }
  }

  const schedule = applyBNSchedule(shuffle(outSchedule, randomizeHash), network, consensusRound)

  return {
    schedule,
    valid_from: schedule[0]?.bn || 0,
    valid_to: schedule[outSchedule.length - 1]?.bn || 1,
    valid_epoch: electionResult.epoch,
  }
}

// networks[this.self.config.get('network.id')]
type NetworkInfo = typeof networks[keyof typeof networks]

type ConsensusRound = {
  nextRoundHeight: number
  pastRoundHeight: number
  currentBlockNumber: number
}

export function calculateConsensusRound(blockNumber: number, network: NetworkInfo): ConsensusRound {
  const { roundLength, totalRounds } = network

  const modLength = roundLength * totalRounds
  const mod3 = blockNumber % modLength
  const pastRoundHeight = blockNumber - mod3

  return {
    nextRoundHeight: blockNumber + (modLength - mod3),
    pastRoundHeight,
    currentBlockNumber: blockNumber,
  }
}

async function getRandomizeHashForBlock(
  { self }: { self: NewCoreService },
  consensusRound: ConsensusRound,
): Promise<string> {
  const { pastRoundHeight } = consensusRound

  //Return block id of previous block to round start.
  //This is used for schedule randomization
  const blockHeader = await self.chainBridge.events.findOne({
    key: pastRoundHeight - 1,
  })

  if (!blockHeader) {
    throw new Error(`could not find hive block header for block: ${pastRoundHeight - 1}`)
  }

  return blockHeader.block_id
}

export type Schedule = ReturnType<typeof weightedSchedule>['schedule']

/**
 * Get block producer schedule
 * @param blockHeight
 * @returns
 */
export async function getBlockSchedule(witness: WitnessServiceV2, blockHeight: number): Promise<Schedule> {
  const { self } = witness
  const network: NetworkInfo = networks[self.config.get('network.id')]
  const consensusRound = calculateConsensusRound(blockHeight, network)
  const electionResult = await self.electionManager.getValidElectionOfblock(blockHeight)
  // console.log(electionResult.epoch, this.witnessSchedule.valid_epoch, consensusRound.randomizeHash, this.witnessSchedule.valid_height)
  if (
    witness.witnessSchedule &&
    witness.witnessSchedule.valid_height === consensusRound.pastRoundHeight &&
    electionResult?.epoch === witness.witnessSchedule.valid_epoch
  ) {
    // console.log('this.witnessSchedule.valid_to', this.witnessSchedule.valid_height, blockHeight, consensusRound)
    return witness.witnessSchedule.schedule
  }

  const [witnessNodes, randomizeHash] = await Promise.all([self.electionManager.getMembersOfBlock(blockHeight),getRandomizeHashForBlock(witness, consensusRound)])

  const { schedule, valid_to, valid_from, valid_epoch } = weightedSchedule(
    network,
    witnessNodes,
    electionResult,
    randomizeHash,
    consensusRound,
  )

  witness.witnessSchedule = {
    schedule,
    valid_epoch,
    valid_height: consensusRound.pastRoundHeight,
  }

  return schedule
}
