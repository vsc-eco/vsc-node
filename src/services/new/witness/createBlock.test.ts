import { createIPFSClient, createMongoDBClient } from "@/utils"
import { WitnessServiceV2 } from "."
import { NewCoreService } from ".."
import { TransactionDbRecordV2 } from "../types"
import { ContractEngineV2 } from "../contractEngineV2"
import { mongo } from "@/services/db"
import { txs } from "@/vsc-new.transaction_pool-part"
import { txsFull } from "@/vsc-new.transaction_pool-full"
import { shuffle } from "./schedule"

const mock = <T extends object>(name: string, handler: ProxyHandler<T>) => new Proxy({} as T, new Proxy(handler, {
        get(target, method, receiver) {
            const errorFn = (_: any, prop: string | symbol) => {
                throw new Error(`${String(method)}ing ${name}.${String(prop)} is not mocked yet`)
            }
            if (method in target) {
                return Reflect.get(target, method, errorFn);
            }
            return errorFn
        },
    }))

describe('createBlock', () => {
    // suppress all console.log's for now
    const log = console.log.bind(console);
    console.log = () => {}

    // mocks
    const db = createMongoDBClient('new')
    const ipfs = createIPFSClient({url: 'http://127.0.0.1:5101'})
    const core: NewCoreService = mock('core', {
        get(target, p, receiver) {
            switch(p) {
                case 'ipfs':
                    return ipfs;
                case 'transactionPool':
                    return mock('transactionPool', {
                        get(target, p, receiver) {
                            if (p === 'txDb') {
                                return db.collection('transaction_pool')
                            }
                            throw new Error('hmm2')
                        },
                    });
                case 'nonceMap':
                    return db.collection('nonce_map');
                case 'contractEngine':
                    return contractEngine;
                case 'db':
                    return db;
                case 'chainBridge':
                    return mock('chainBridge', {
                        get(target, p, receiver) {
                            switch (p) {
                                case 'streamParser':
                                    return mock('streamParser', {
                                        get(target, p, receiver) {
                                            switch (p) {
                                                case 'addParser':
                                                    return () => {}
                                            }
                                        },
                                    })
                                case 'blockHeaders':
                                    return db.collection('block_headers');
                            }
                        },
                    })
            }
            throw new Error('hmm')
        },
    })
    const contractEngine = new ContractEngineV2(core)

    beforeAll(async () => {
        await mongo.connect();
        await contractEngine.init();
    })

    afterAll(async () => {
        await mongo.close()
        setTimeout(() => process.exit(0), 1000)
    })

    // service to test
    let witness: WitnessServiceV2
    beforeEach(async () => {
        witness = new WitnessServiceV2(core)
        witness.blockHeaders = db.collection('block_headers')
    })

    const testCases: { inputName: string; txs: TransactionDbRecordV2[] }[] = [
        {
            inputName: 'no input',
            txs: []
        },
        {
            inputName: 'many contract calls',
            // @ts-ignore
            txs,
        },
        {
            inputName: 'very many contract calls',
            // @ts-ignore
            txs: txsFull,
        }
    ]

    for (const {inputName, txs} of testCases) {

        it(`should create consistent output for the same input with ${inputName}`, async () => {
            const createBlock = () => witness.createBlock({
                start_height: 86407310 + 1,
                end_height: 86424500,
                offchainTxsUnfiltered: txs
            })

            const [b1, b2] = await Promise.all([createBlock(), createBlock()])

            expect(b1).toStrictEqual(b2)
            expect(b1.rawData.txs.length - txs.length).toMatchSnapshot()
            expect(b1).toMatchSnapshot()
        })

        it(`should create differ output for the same but jumbled input with ${inputName}`, async () => {
            let count = 0;
            const createBlock = () => witness.createBlock({
                start_height: 86407310 + 1,
                end_height: 86424500,
                offchainTxsUnfiltered: shuffle(txs, (++count).toString())
            })

            const [b1, b2] = await Promise.all([createBlock(), createBlock()])

            if (txs.length === 0) { 
                expect(b1).toStrictEqual(b2)
            } else {
                expect(b1).not.toStrictEqual(b2)
            }
            expect(b1.rawData.txs.length - txs.length).toMatchSnapshot()
            expect(b2.rawData.txs.length - txs.length).toMatchSnapshot()
            expect(b1).toMatchSnapshot()
            expect(b2).toMatchSnapshot()
        })

    }
})