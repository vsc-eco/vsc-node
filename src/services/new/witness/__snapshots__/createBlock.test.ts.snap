// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBlock should create consistent output for the same input with many contract calls 1`] = `-254`;

exports[`createBlock should create consistent output for the same input with many contract calls 2`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "TEh3mHQedmhtKy-GnMjv09uikPLzX4Z2yrkZycDWxKU",
    "sig_root": "Yy-kozWLKgJrPdZHzgRnxG9Y2CguMLMfJX2lXjsZg54",
    "txs": [
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create consistent output for the same input with no input 1`] = `1`;

exports[`createBlock should create consistent output for the same input with no input 2`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "y6uZ4_wbTETEvKlGKXBsb0IXDXPdvigMw4nOn95pXsA",
    "sig_root": null,
    "txs": [
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create consistent output for the same input with very many contract calls 1`] = `-1728`;

exports[`createBlock should create consistent output for the same input with very many contract calls 2`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "6hPG0yYU8MUkpiEiPgPlWPItafzCLBr7KNAUMhr5iTA",
    "sig_root": "voZM2n8VDHm1GK5xsb9BO8GMOyXnke8_xneg8FAL6Ew",
    "txs": [
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicxqtgf7wz3d6dxzju5xqrcfaau3y3d5uqmi6tsmua2t6ghxtwftq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreihrgoc3uingpntmlbcvwta7q7h7cppeykeeg3ld6u5h2bkuhr6lvq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiglk2mgiqhsffutp23ym2bhlkfeinpkvy265iv6vhas5hm4gigz4e",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicbb2jo44uzahr2e67wet45trjngc6kmhqdbmmumcgoasujywlpke",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiau63ookeim7oejypmtl33xmp3vbyumjbj5fhgalfjifmxz67bvn4",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreido725sh37l6su73c7ly62vtfhzxvnr6tlnds3nox553loibwcs2m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreidnzz52qqojy4mzjnykkteakmhdlg55yg4uv7qypkosliz4i6mp6a",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreifq2evbxt2xs5fxldedv6mdye6fyurl2exfpwundbb7d3bdqwwrty",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicbf4il2wzjvr2pxac6zas3qigszgt2ycnmhp62bhwfl4dppz5gjq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreieghdlx266eqkqtm245qc5bpu6xbxzmu32d36x3shhgdvnfqs7hnu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiap4fmnth3sosv6rr4yay3t3mprjagzst7auvsp5w3jkhojny7p2u",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreid2qsixgkyzb6q4pbvpxghnsfngmv2nc4kdwb7l7sa3qaumjcvg4m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreictg2v7jlhkxuytyf27np6lksltg2l5u6ap7ymhvuiyqgs5ba4bei",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreignqw7fbqqq6wxqvr5bdwkgp526ek7d6vx47vkmopk4cfjd6ldpii",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with many contract calls 1`] = `-254`;

exports[`createBlock should create differ output for the same but jumbled input with many contract calls 2`] = `-253`;

exports[`createBlock should create differ output for the same but jumbled input with many contract calls 3`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "TEh3mHQedmhtKy-GnMjv09uikPLzX4Z2yrkZycDWxKU",
    "sig_root": "Yy-kozWLKgJrPdZHzgRnxG9Y2CguMLMfJX2lXjsZg54",
    "txs": [
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with many contract calls 4`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "Ln0CejY-X4JvP0Dn0HFuan_baX-JeP3yJujvkehaVUU",
    "sig_root": "ygaIDkVDW3a8cOKpA9EHFKSf5duFgsEokYpOoeLWSqQ",
    "txs": [
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreia6ljwjlqpjlrb5ojllalgekgck6mz3s3l7w6ssz5s6x2tjljujcu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with no input 1`] = `1`;

exports[`createBlock should create differ output for the same but jumbled input with no input 2`] = `1`;

exports[`createBlock should create differ output for the same but jumbled input with no input 3`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "y6uZ4_wbTETEvKlGKXBsb0IXDXPdvigMw4nOn95pXsA",
    "sig_root": null,
    "txs": [
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with no input 4`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "y6uZ4_wbTETEvKlGKXBsb0IXDXPdvigMw4nOn95pXsA",
    "sig_root": null,
    "txs": [
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with very many contract calls 1`] = `-1723`;

exports[`createBlock should create differ output for the same but jumbled input with very many contract calls 2`] = `-1726`;

exports[`createBlock should create differ output for the same but jumbled input with very many contract calls 3`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "IWjlzL26LVrL29FjXs_zel-WGOc0Je_2Y3WTVIsBi1s",
    "sig_root": "-ptu05jLgsiZkRBhVbnRJBQYWHkhuThWoQYta--_Xao",
    "txs": [
      {
        "id": "bafyreid2qsixgkyzb6q4pbvpxghnsfngmv2nc4kdwb7l7sa3qaumjcvg4m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiglk2mgiqhsffutp23ym2bhlkfeinpkvy265iv6vhas5hm4gigz4e",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreido725sh37l6su73c7ly62vtfhzxvnr6tlnds3nox553loibwcs2m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreifo2caumdy2wnpwbbubyxdx4ordsok6jmtu4rvsegpnotczdshbd4",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicxqtgf7wz3d6dxzju5xqrcfaau3y3d5uqmi6tsmua2t6ghxtwftq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreieghdlx266eqkqtm245qc5bpu6xbxzmu32d36x3shhgdvnfqs7hnu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiagaam5wkkmooo5s2jpp6baxcnefuaewydg32f5uhpfqsx72brotu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreignqw7fbqqq6wxqvr5bdwkgp526ek7d6vx47vkmopk4cfjd6ldpii",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreidnzz52qqojy4mzjnykkteakmhdlg55yg4uv7qypkosliz4i6mp6a",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicbf4il2wzjvr2pxac6zas3qigszgt2ycnmhp62bhwfl4dppz5gjq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiau63ookeim7oejypmtl33xmp3vbyumjbj5fhgalfjifmxz67bvn4",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreifq2evbxt2xs5fxldedv6mdye6fyurl2exfpwundbb7d3bdqwwrty",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiap4fmnth3sosv6rr4yay3t3mprjagzst7auvsp5w3jkhojny7p2u",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicj6uzqw4xhb6geev6k5ax4n3n7pduwmqh5wfxisubptuoqzn7jiy",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreigcyjl43mvuecmfn7xtsh6fsdxzboanjsnkbyh7xnqf4vusy6meye",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreictg2v7jlhkxuytyf27np6lksltg2l5u6ap7ymhvuiyqgs5ba4bei",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiblwknqy6spdfmkoe7x3m2alyqaapi62z2drgskisr6twp7rqnw4m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreihrgoc3uingpntmlbcvwta7q7h7cppeykeeg3ld6u5h2bkuhr6lvq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicbb2jo44uzahr2e67wet45trjngc6kmhqdbmmumcgoasujywlpke",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;

exports[`createBlock should create differ output for the same but jumbled input with very many contract calls 4`] = `
BlockContainer {
  "rawData": {
    "__t": "vsc-block",
    "__v": "0.1",
    "headers": {
      "prevb": "bafyreifndgmch5ex2dmgoltnq7nn2mrtjjsvbicbcfzgemesomdem2jgte",
    },
    "merkle_root": "96E2FWVgRCixyEY57MDw6Xa32vzXll2XEwk3fn9ZsJA",
    "sig_root": "1yfhw-XxRZ7boGpcZX9ltriJdqjgfOwBUU2IHCSq3V4",
    "txs": [
      {
        "id": "bafyreiglk2mgiqhsffutp23ym2bhlkfeinpkvy265iv6vhas5hm4gigz4e",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreid2qsixgkyzb6q4pbvpxghnsfngmv2nc4kdwb7l7sa3qaumjcvg4m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreihrgoc3uingpntmlbcvwta7q7h7cppeykeeg3ld6u5h2bkuhr6lvq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiap4fmnth3sosv6rr4yay3t3mprjagzst7auvsp5w3jkhojny7p2u",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreidz6lgett5eiq4j7tdk74wf4h5fgzqgutdz2xchhxpsvarnmdjkey",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicgfn4kkkdaovz3dmjsjoaf2iywoyuym7uv2v7q4v4qlemwlw5n5u",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreiau63ookeim7oejypmtl33xmp3vbyumjbj5fhgalfjifmxz67bvn4",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicljav3fbfge6crbpox43vngfksbapxupajye5bns27k442yrtsjm",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicxqtgf7wz3d6dxzju5xqrcfaau3y3d5uqmi6tsmua2t6ghxtwftq",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreicbb2jo44uzahr2e67wet45trjngc6kmhqdbmmumcgoasujywlpke",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreihmtgg7snwhbts2d7gj5ia5codc2fysfpo5xiqfecai3hh7zabeje",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreidnzz52qqojy4mzjnykkteakmhdlg55yg4uv7qypkosliz4i6mp6a",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreido725sh37l6su73c7ly62vtfhzxvnr6tlnds3nox553loibwcs2m",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreignqw7fbqqq6wxqvr5bdwkgp526ek7d6vx47vkmopk4cfjd6ldpii",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreia6ljwjlqpjlrb5ojllalgekgck6mz3s3l7w6ssz5s6x2tjljujcu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreigyebux25pfac24mzuxb53efhbsd5l5pshexd4m76tifdjudecegi",
        "op": "call_contract",
        "type": 1,
      },
      {
        "id": "bafyreieghdlx266eqkqtm245qc5bpu6xbxzmu32d36x3shhgdvnfqs7hnu",
        "op": "call_contract",
        "type": 1,
      },
      {
        "contract_id": "vs41q9c3ygynfp6kl86qnlaswuwvam748s5lvugns5schg4hte5vhusnx7sg5u8falrt",
        "id": "bafyreib4m744jssm2bt5bqpvz27567orqfljnicrrvjb6rocldd55ogkwu",
        "type": 2,
      },
    ],
  },
  "ref_end": 86424500,
  "ref_start": 86407311,
}
`;
